How many late days are you using for this assignment? 

In [2]:
# Imports and pip installations (if needed)
#no late day

# Part 1: Load the dataset

In [8]:
# Load the given datasets
import pandas as pd
from collections import Counter

df = pd.read_csv("C:/Users/<redacted>/Desktop/Assignment 3/assignment-3-data-cleaning-and-preprocessing-<redacted>/chronic_kidney_disease_numerical.csv")

print(df)
# Print the data


     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0   
2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0   
3       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...   
415     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0   
416     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0   
417     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0   
418     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0   
419     926033  50.0   70.0  230.0   50.0  2.2    NaN  NaN  12.0  41.0   

        wbcc  rbcc  
0     7300.0   6.4  
1     7200.0   5.5  
2     5500.0   5.7  
3     8200.0   4.5  
4     

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)




Answer:
Features: These are the input variables used to predict an outcome. In this dataset, features likely include "age," "bp" (blood pressure), "bgr" (blood glucose level), "bu" (blood urea), "sc" (serum creatinine), "sod" (sodium), "pot" (potassium), "hemo" (hemoglobin), "pcv" (packed cell volume), "wbcc" (white blood cell count), and "rbcc" (red blood cell count).
Labels: The dataset does not explicitly state the label, but in medical datasets, it is often a diagnosis or disease classification (e.g., presence of kidney disease, diabetes, or other conditions).
The dataset consists mostly of numerical features, as all the visible columns contain continuous numerical values.
There do not seem to be any categorical features in the displayed portion, but there could be in the full dataset (e.g., categorical variables for "gender" or "disease presence").
Examples of Numerical and Categorical Columns:
Numerical Columns (Continuous values)
Age – Represents the age of the patient.
Blood Pressure (bp) – Measures blood pressure levels.
Hemoglobin (hemo) – A measure of hemoglobin in the blood.
Categorical Columns (if they exist in the full dataset)
Gender – If present, it would have categories like "Male" and "Female."
Presence of Disease – If the dataset includes a diagnosis column, it may have values like "Yes" or "No."
Blood Group – If available, it might have categories such as "A," "B," "AB," or "O."


# Part 3: Data Preprocessing




## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [10]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
duplicate_rows = df[df.duplicated(subset=['unique_id'], keep=False)]
print(duplicate_rows)
len(duplicate_rows)
# Drop these duplicate rows
df_clean = df.drop_duplicates(subset=['unique_id'], keep="first")

df_clean.to_csv("numerical_dataset.csv", index=False)

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
df = pd.read_csv("C:/Users/<redacted>/Desktop/Assignment 3/assignment-3-data-cleaning-and-preprocessing-<redacted>/chronic_kidney_disease_categorical.csv")
duplicate_rows = df[df.duplicated(subset=['unique_id'], keep=False)]

len(duplicate_rows)
df_clean = df.drop_duplicates(subset=['unique_id'], keep="first")

df_clean.to_csv("categorical_dataset.csv", index=False)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN   
15      224481  54.0   60.0  125.0   21.0  1.3  137.0  3.4  15.0  46.0   
35      431871  57.0   60.0  132.0   18.0  1.1  150.0  4.7  15.4  42.0   
42      441170  79.0   80.0  111.0   44.0  1.2  146.0  3.6  16.3  40.0   
53      261137  35.0   60.0  105.0   39.0  0.5  135.0  3.9  14.7  43.0   
54      721270  58.0   70.0  102.0   48.0  1.2  139.0  4.3  15.0  40.0   
61      872034  70.0   90.0  184.0   98.6  3.3  138.0  3.9   5.8   NaN   
78      501321   4.0    NaN   99.0   23.0  0.6  138.0  4.4  12.0  34.0   
79      283427  55.0   80.0  140.0   49.0  0.5  150.0  4.9  15.7  47.0   
81      302906  59.0  100.0    NaN   96.0  6.4    NaN  NaN   6.6   NaN   
98      708255   3.0    NaN    NaN   22.0  0.7    NaN  NaN  10.7  34.0   
107     872034  70.0   90.0  184.0   9

## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

No, there are 10 unique ids in categorical dataset that are different from numerical dataset
IDs present only in categorical dataset: {208034, 240430, 488947, 325238, 254775, 292312, 298297, 663387, 948317, 266335}


Answer:

In [11]:
categorical_dataset = pd.read_csv("categorical_dataset.csv")
numerical_dataset = pd.read_csv("numerical_dataset.csv")

In [12]:
ids_categorical = set(categorical_dataset['unique_id'])
ids_numerical = set(numerical_dataset['unique_id'])

common_ids = ids_categorical.intersection(ids_numerical)
only_in_categorical = ids_categorical - ids_numerical
only_in_numerical = ids_numerical - ids_categorical

print(f"Common unique IDs in both datasets: {len(common_ids)}")
print(f"Unique ids only in categorical dataset: {len(only_in_categorical)}")
print(f"Unique ids only in numerical dataset: {len(only_in_numerical)}")

Common unique IDs in both datasets: 390
Unique ids only in categorical dataset: 10
Unique ids only in numerical dataset: 0


In [13]:
if only_in_categorical:
    print("IDs present only in categorical dataset:", only_in_categorical)
if only_in_numerical:
    print("IDs present only in numerical dataset", only_in_numerical)

IDs present only in categorical dataset: {208034, 240430, 488947, 325238, 254775, 292312, 298297, 663387, 948317, 266335}


In [14]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
merged_inner = pd.merge(categorical_dataset, numerical_dataset, on="unique_id", how="inner")
print(merged_inner.head())
#Print the combined dataset


   unique_id   al   su       rbc        pc         pcc          ba  htn   dm  \
0     224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes  yes   
1     992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   no   
2     308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   no   
3     450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   no   
4     881763  0.0  0.0    normal    normal  notpresent  notpresent   no   no   

  cad  ...    bp    bgr    bu   sc    sod  pot  hemo   pcv    wbcc  rbcc  
0  no  ...  60.0  125.0  21.0  1.3  137.0  3.4  15.0  46.0     NaN   NaN  
1  no  ...  70.0   82.0  36.0  1.1  150.0  3.5  14.5  52.0  9400.0   6.1  
2  no  ...  70.0  139.0  29.0  1.0    NaN  NaN  10.5  32.0     NaN   NaN  
3  no  ...   NaN    NaN  16.0  0.7  138.0  3.2   8.1   NaN     NaN   NaN  
4  no  ...  70.0   75.0  31.0  1.2  141.0  3.5  16.5  54.0  7800.0   6.2  

[5 rows x 25 columns]


In [15]:
merged_inner.to_csv("merged_dataset.csv", index=False)


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [16]:
# Calculate the percentage of rows that contain atleast one missing value
df = pd.read_csv("merged_dataset.csv")
missing_rows = df.isnull().any(axis=1).sum()

total_rows = df.shape[0]
percentage_missing = (missing_rows / total_rows) * 100

print(missing_rows)
print(total_rows)
print(f"Percentage of rows with at least one missing value: {percentage_missing:.2f}%")
# Print %
237
390
#Percentage of rows with at least one missing value: 60.77%
# Drop these rows from the dataset
df_cleaned = df.dropna()

new_total_rows = df_cleaned.shape[0]
print(new_total_rows)
# Print the Datase

237
390
Percentage of rows with at least one missing value: 60.77%
153


In [18]:
df_cleaned.to_csv("drop_na_dataset.csv", index=False)

In [19]:
df = pd.read_csv("drop_na_dataset.csv")

## Part 3.4: Sort the dataset according to the Labels

In [20]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
if 'Target' in df.columns:
    df_cleaned = df.sort_values(by="Target").reset_index(drop=True)

    print(df_cleaned.head())

    df_cleaned.to_csv("sorted_dataset.csv", index=False)
# Print the dataset

   unique_id   al   su       rbc        pc         pcc          ba  htn   dm  \
0     534245  2.0  0.0    normal  abnormal     present  notpresent   no   no   
1     349892  1.0  3.0  abnormal  abnormal  notpresent  notpresent  yes  yes   
2     343710  2.0  0.0  abnormal  abnormal  notpresent  notpresent  yes  yes   
3     102677  3.0  0.0    normal  abnormal  notpresent  notpresent  yes   no   
4     514721  2.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   no   

   cad  ...    bp    bgr     bu    sc    sod  pot  hemo   pcv     wbcc  rbcc  
0   no  ...  70.0  117.0   52.0   2.2  136.0  3.8  10.0  30.0  19100.0   3.7  
1  yes  ...  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  10200.0   4.1  
2  yes  ...  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0   9200.0   3.2  
3   no  ...  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0   3800.0   2.1  
4   no  ...  90.0  129.0  107.0   6.7  131.0  4.8   9.1  29.0   6400.0   3.4  

[5 rows x 25 columns]


In [21]:
df = pd.read_csv("sorted_dataset.csv")

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [22]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
existing_categorical_columns = [col for col in categorical_columns if col in df.columns]
df_encoded = pd.get_dummies(df, columns=existing_categorical_columns, drop_first=False)
print("Encoded Dataset:\n", df_encoded.head())

df_encoded.to_csv("encoded_dataset.csv", index=False)
print("Encoded dataset saved as 'encoded_dataset.csv'.")
# Print the dataset

Encoded Dataset:
    unique_id   al   su   age    bp    bgr     bu    sc    sod  pot  ...  \
0     534245  2.0  0.0  45.0  70.0  117.0   52.0   2.2  136.0  3.8  ...   
1     349892  1.0  3.0  59.0  70.0  424.0   55.0   1.7  138.0  4.5  ...   
2     343710  2.0  0.0  61.0  80.0  173.0  148.0   3.9  135.0  5.2  ...   
3     102677  3.0  0.0  59.0  70.0   76.0  186.0  15.0  135.0  7.6  ...   
4     514721  2.0  0.0  56.0  90.0  129.0  107.0   6.7  131.0  4.8  ...   

   cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_yes  \
0    True    False        True       False   True   False    True    False   
1   False     True        True       False   True   False    True    False   
2   False     True       False        True  False    True   False     True   
3    True    False       False        True  False    True   False     True   
4    True    False        True       False   True   False    True    False   

   Target_ckd  Target_notckd  
0        True          False  


In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: 

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [23]:
df = pd.read_csv("encoded_dataset.csv")


In [24]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
existing_numerical_columns = [col for col in numerical_columns if col in df.columns]

# function to remove outliers
def remove_outliers(df, columns):
    for col in columns:
        mean = df[col].mean()
        std = df[col].std()
        df = df[(df[col] >= mean - 3 * std) & (df[col] <= mean + 3 * std)]
    return df

df_no_outliers = remove_outliers(df, existing_numerical_columns)

print("Dataset after removing outliers:\n", df_no_outliers.head())
df_no_outliers.to_csv("cleaned_no_outliers_dataset.csv", index=False)
print("Dataset without outliers saved as 'cleaned_no_outliers_dataset.csv'.")
# Print the dataset

Dataset after removing outliers:
     unique_id   al   su   age    bp    bgr     bu   sc    sod  pot  ...  \
2      343710  2.0  0.0  61.0  80.0  173.0  148.0  3.9  135.0  5.2  ...   
4      514721  2.0  0.0  56.0  90.0  129.0  107.0  6.7  131.0  4.8  ...   
7      124304  3.0  4.0  69.0  70.0  264.0   87.0  2.7  130.0  4.0  ...   
8      980291  3.0  0.0  62.0  70.0  122.0   42.0  1.7  136.0  4.7  ...   
10     615697  4.0  0.0   6.0  60.0   94.0   67.0  1.0  135.0  4.9  ...   

    cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_yes  \
2    False     True       False        True  False    True   False     True   
4     True    False        True       False   True   False    True    False   
7    False     True        True       False  False    True    True    False   
8     True    False        True       False   True   False    True    False   
10    True    False       False        True   True   False    True    False   

    Target_ckd  Target_notckd  
2       

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [31]:
# Normalize the all Numerical Attributes in the dataset.
from sklearn.preprocessing import MinMaxScaler
df = pd.read_csv("cleaned_no_outliers_dataset.csv")
numerical_columns = ['al', 'su', 'age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

scaler = MinMaxScaler()

df.scaled = df.copy()
df.scaled[numerical_columns] = scaler.fit_transform(df[numerical_columns])
print("Normalized Dataset:\n", df.scaled.head())
# Print the dataset

Normalized Dataset:
    unique_id    al   su       age    bp       bgr        bu        sc  \
0     343710  0.50  0.0  0.743243  0.50  0.442060  0.901961  0.432099   
1     514721  0.50  0.0  0.675676  0.75  0.253219  0.633987  0.777778   
2     124304  0.75  1.0  0.851351  0.25  0.832618  0.503268  0.283951   
3     980291  0.75  0.0  0.756757  0.25  0.223176  0.209150  0.160494   
4     615697  1.00  0.0  0.000000  0.00  0.103004  0.372549  0.074074   

        sod       pot  ...  cad_no  cad_yes  appet_good  appet_poor  pe_no  \
0  0.500000  0.793103  ...   False     True       False        True  False   
1  0.366667  0.655172  ...    True    False        True       False   True   
2  0.333333  0.379310  ...   False     True        True       False  False   
3  0.533333  0.620690  ...    True    False        True       False   True   
4  0.500000  0.689655  ...    True    False       False        True   True   

   pe_yes  ane_no  ane_yes  Target_ckd  Target_notckd  
0    True   Fal

C:\Users\nampe\AppData\Local\Temp\ipykernel_10132\4109414889.py:8: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
  df.scaled = df.copy()


In [32]:
df.scaled.to_csv("normalized_dataset.csv", index=False)
print("Normalized dataset saved as 'normalized_dataset.csv'.")

Normalized dataset saved as 'normalized_dataset.csv'.


## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:

In [34]:
#Remove that column
df = pd.read_csv("normalized_dataset.csv")

columns_to_drop = ['unique_id']

df_cleaned = df.drop(columns=columns_to_drop)

print("Dataset after removing unnecessary columns:\n", df_cleaned.head())


# Print the dataset

Dataset after removing unnecessary columns:
      al   su       age    bp       bgr        bu        sc       sod  \
0  0.50  0.0  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
1  0.50  0.0  0.675676  0.75  0.253219  0.633987  0.777778  0.366667   
2  0.75  1.0  0.851351  0.25  0.832618  0.503268  0.283951  0.333333   
3  0.75  0.0  0.756757  0.25  0.223176  0.209150  0.160494  0.533333   
4  1.00  0.0  0.000000  0.00  0.103004  0.372549  0.074074  0.500000   

        pot      hemo  ...  cad_no  cad_yes  appet_good  appet_poor  pe_no  \
0  0.793103  0.000000  ...   False     True       False        True  False   
1  0.655172  0.138614  ...    True    False        True       False   True   
2  0.379310  0.475248  ...   False     True        True       False  False   
3  0.620690  0.485149  ...    True    False        True       False   True   
4  0.689655  0.217822  ...    True    False       False        True   True   

   pe_yes  ane_no  ane_yes  Target_ckd  Target_notckd

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [36]:
# Export the dataframe to a new csv file
df_cleaned.to_csv("final_dataset.csv", index=False)
print("Final cleaned dataset saved as 'final_dataset.csv'.")
# Export the dataframe to a new json file
json_file_path = "final_dataset.json"
df_cleaned.to_json(json_file_path, orient="records", indent=4)
print(f"Final dataset saved as JSON: '{json_file_path}'.")

Final cleaned dataset saved as 'final_dataset.csv'.
Final dataset saved as JSON: 'final_dataset.json'.


# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [37]:
top_15_rows = df_cleaned.head(15)
print(top_15_rows)

      al    su       age    bp       bgr        bu        sc       sod  \
0   0.50  0.00  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
1   0.50  0.00  0.675676  0.75  0.253219  0.633987  0.777778  0.366667   
2   0.75  1.00  0.851351  0.25  0.832618  0.503268  0.283951  0.333333   
3   0.75  0.00  0.756757  0.25  0.223176  0.209150  0.160494  0.533333   
4   1.00  0.00  0.000000  0.00  0.103004  0.372549  0.074074  0.500000   
5   0.75  0.25  0.662162  0.50  0.618026  0.411765  0.432099  0.566667   
6   0.25  0.00  0.540541  0.00  0.399142  0.535948  0.358025  0.700000   
7   1.00  0.50  0.783784  1.00  0.399142  0.287582  0.839506  0.666667   
8   1.00  0.00  0.567568  0.50  0.107296  1.000000  0.901235  0.533333   
9   1.00  0.75  0.851351  0.25  0.618026  0.562092  0.728395  0.000000   
10  0.75  0.50  0.905405  1.00  0.965665  0.522876  0.641975  0.666667   
11  1.00  0.00  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
12  1.00  0.25  0.783784  0.00  0.7253

Paste here: 

| al  | su  | age     | bp  | bgr     | bu      | sc      | sod     | pot     | hemo    | ... | cad_no | cad_yes | appet_good | appet_poor | pe_no |
|-----|-----|---------|-----|---------|---------|---------|---------|---------|---------|-----|--------|---------|------------|------------|------|
| 0.50 | 0.0 | 0.734234 | 0.50 | 0.442060 | 0.901961 | 0.432099 | 0.500000 | 0.793103 | 0.000000 | ... | False  | True    | True       | False      | False |
| 0.50 | 0.0 | 0.675676 | 0.75 | 0.253219 | 0.633987 | 0.777778 | 0.366667 | 0.655172 | 0.138614 | ... | True   | False   | True       | True       | False |
| 0.75 | 0.0 | 0.851351 | 0.25 | 0.223176 | 0.209150 | 0.160493 | 0.533333 | 0.379310 | 0.475248 | ... | True   | False   | True       | False      | False |
| 0.75 | 0.0 | 0.756757 | 0.25 | 0.223176 | 0.209150 | 0.160493 | 0.533333 | 0.620690 | 0.485149 | ... | True   | False   | True       | False      | False |
| 1.00 | 0.0 | 0.000000 | 0.0  | 0.108034 | 0.372549 | 0.074074 | 0.500000 | 0.689655 | 0.217822 | ... | True   | False   | True       | False      | False |
| 0.25 | 0.0 | 0.540541 | 0.50 | 0.399142 | 0.535948 | 0.358025 | 0.700000 | 0.689655 | 0.316832 | ... | True   | False   | True       | False      | False |
| 1.00 | 0.0 | 0.567568 | 0.50 | 0.107296 | 1.000000 | 0.921235 | 0.533333 | 0.379310 | 0.207921 | ... | True   | False   | True       | False      | True  |
| 0.75 | 0.0 | 0.851351 | 0.25 | 0.618026 | 0.562092 | 0.723895 | 0.500000 | ... | False  | False  | False      | False      | False |
| 0.75 | 0.0 | 0.904595 | 0.25 | 0.965665 | 0.522876 | 0.641975 | 0.166667 | ... | False  | True   | False      | False      | False |
| 1.00 | 0.0 | 0.202703 | 0.75 | 0.157898 | 0.196078 | 0.160494 | 0.166667 | ... | False  | False  | False      | False      | False |
| 1.00 | 0.0 | 0.783784 | 0.0  | 0.723522 | 0.313725 | 0.481481 | 0.566667 | ... | False  | True   | False      | False      | False |
| 1.00 | 0.0 | 0.25     | 0.676676 | 0.25 | 0.600838 | 0.104575 | 0.160494 | ... | False  | True   | False      | False      | True  |
| 1.00 | 0.0 | 0.567568 | 0.50 | 0.270386 | 0.843137 | 1.000000 | 0.400000 | ... | True   | True   | False      | False      | False |


### Paste the markdown table here

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [ ]:
column_mapping = {
    'age': 'Age',
    'bp': 'Blood Pressure',
    'sg': 'Specific Gravity',
    'al': 'Albumin',
    'su': 'Sugar',
    'rbc': 'Red Blood Cells',
    'pc': 'Pus Cell',
    'pcc': 'Pus Cell Clumps',
    'ba': 'Bacteria',
    'bgr': 'Blood Glucose Random',
    'bu': 'Blood Urea',
    'sc': 'Serum Creatinine',
    'sod': 'Sodium',
    'pot': 'Potassium',
    'hemo': 'Hemoglobin',
    'pcv': 'Packed Cell Volume',
    'wbcc': 'White Blood Cell Count',
    'rbcc': 'Red Blood Cell Count',
    'htn': 'Hypertension',
    'dm': 'Diabetes Mellitus',
    'cad': 'Coronary Artery Disease',
    'appet': 'Appetite',
    'pe': 'Pedal Edema',
    'ane': 'Anemia'
}


In [ ]:
# Code to rename all the columns in the dataset
import pandas as pd

# Load your dataset (adjust the filename as needed)
df = pd.read_csv("cleaned_no_outliers_dataset.csv")

# Rename the columns
df.rename(columns=column_mapping, inplace=True)

# Display the first few rows to verify changes
print(df.head())

# Save the updated dataset (optional)
df.to_csv("renamed_dataset.csv", index=False)


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [ ]:
# Converted SQL to Pandas code
import pandas as pd

# Load your dataset (change file name as needed)
df = pd.read_csv("your_dataset.csv")  

# Perform the group by operation
target_counts = df.groupby("Target").size().reset_index(name="count")

# Display the result
print(target_counts)


*Describe what the above code does here*
This counts the occurrences of each unique value in the "Target" column and displays the result as a new DataFrame with two columns:

Target (Unique categories in the column) count (Number of occurrences for each category)